text/0000-infer-outlives.md

```diff
@@ -0,0 +1,575 @@
+where the lifetime and the type are combined within the value of an
+associated type, but not in one of the impl's input types. For
+example:
+
+```
+trait MakeRef<'a> {
+  type Type;
+}
+
+impl<'a, T> MakeRef<'a> for Vec<T>
+  where T: 'a // still required
+{
+  type Type = &'a T;
+}
+```
+
+In this case, the impl has two inputs -- the lifetime `'a` and the
+type `Vec<T>` (note that `'a` and `T` are the impl parameters; the
+inputs come from the parameters of the trait that is being
+implemented). Neither of these inputs requires that `T: 'a`, so when
+we try to specify the value of the associated type as `&'a T`, we
+still require a where clause to infer that `T: 'a` must hold.
+
+In turn, if this associated type were used in a struct, where-clauses
+would be required. As we'll see in the reference-level explanation,
+this is a consequence of the fact that we do inference without regard
+for associated type normalization, but it makes for a relatively
+simple rule -- explicit where clauses are needed in the preseence of
+impls like the one above:
+
+```rust
+struct Foo<'a, T>
+  where T: 'a // still required, not inferred from `field`
+{
+  field: <Vec<T> as MakeRef<'a>>::Type
+}    
+```
+
+As the algorithm is currently framed, outlives requirements written on
+traits must also be explicitly propagated; however, this will typically
+occur as part of the existing bounds:
+
+```rust
+trait Trait<'a> where Self: 'a {
+  type Type;
+}
+
+struct Foo<'a, T>
+  where T: Trait<'a> // implies `T: 'a` already, so no error
+{
+  r: <T as Trait<'a>>::Type // requires that `T: 'a` to be WF
+}
+```
+
+### Trait object lifetime defaults
+
+[RFC 599][] (later amended by [RFC 1156]) specified the defaulting
+rules for trait object types. Typically, a trait object type that
+appears as a parameter to a struct is given the implicit bound
+`'static`; hence `Box<Debug>` defaults to `Box<Debug +
+'static>`. References to trait objects, however, are given by default
+the lifetime of the reference; hence `&'a Debug` defaults to `&'a
+(Debug + 'a)`.
+
+Structs that contain explicit `T: 'a` where-clauses, however, use the
+default given lifetime `'a` as the default for trait objects.
+Therefore, given a struct definition like the following:
+
+```rust
+struct Ref<'a, T> where T: 'a + ?Sized { .. }
+```
+
+The type `Ref<'x, Debug>` defaults to `Ref<'x, Debug + 'x>` and not
+`Ref<'x, Debug + 'static>`. Effectively the `where T: 'a` declaration
+acts as a kind of signal that `Ref` acts as a "reference to `T`".
+
+This RFC does not change these defaulting rules: in particular, these
+defaults are applied **before** where-clause inference takes place,
+and hence are not affected by the results. Trait object defaulting
+therefore requires an explicit `where T: 'a` declaration on the
+struct; in fact, such explicit declarations can be thought of as
+existing primarily for the purpose of informing trait object lifetime
+defaults, since they are typically not needed otherwise.
```

Comment:
Is my understanding correct that it would be a backwards-compatible change to add an explicit `where` clause, but that removing a `where` clause (and leaving it to be inferred) would be a breaking change? In the first case, you're providing extra information to trait object lifetime defaulting, so it seems "better" from a user's perspective. If a library author sees that users are having to frequently specify the extra lifetime bound, they may want to add a `where` clause in order to enable lifetime defaulting.